File: apps/clojure/design-pattern/src/design_pattern/composite.clj
(ns design-pattern.composite)

;; Composite パターン
;; ネストされたマップとベクタで木構造を表現する。
;; マルチメソッドで :type に応じたディスパッチを行う。

(defn leaf
  "リーフノード（タスク）を作成する。"
  [name duration]
  {:type :leaf :name name :duration duration})

(defn composite
  "コンポジットノード（プロジェクト）を作成する。"
  [name & children]
  {:type :composite :name name :children (vec children)})

(defn add-child
  "コンポジットに子ノードを追加する。"
  [parent child]
  (update parent :children conj child))

(defn remove-child
  "コンポジットから指定名の子ノードを削除する。"
  [parent child-name]
  (update parent :children (fn [cs] (vec (remove #(= (:name %) child-name) cs)))))

;; --- マルチメソッドによるディスパッチ ---

(defmulti total-duration
  "ノードの合計所要時間を計算する。"
  :type)

(defmethod total-duration :leaf [node]
  (:duration node))

(defmethod total-duration :composite [node]
  (reduce + 0 (map total-duration (:children node))))

(defmulti display
  "ノードをインデント付きで表示する。"
  (fn [node _depth] (:type node)))

(defmethod display :leaf [node depth]
  (str (apply str (repeat depth "  ")) "- " (:name node)
       " (" (:duration node) "h)\n"))

(defmethod display :composite [node depth]
  (str (apply str (repeat depth "  ")) "+ " (:name node) "\n"
       (apply str (map #(display % (inc depth)) (:children node)))))

File: apps/elixir/design-pattern/lib/design_pattern/composite.ex
defmodule DesignPattern.Composite do
  @moduledoc """
  Composite パターン: 再帰的データ構造とパターンマッチングによる木構造の表現。
  タグ付きタプルでリーフとコンポジットを区別する。
  """

  @doc "リーフノードを作成する"
  def leaf(name, value) do
    {:leaf, name, value}
  end

  @doc "コンポジットノードを作成する"
  def composite(name, children \\ []) do
    {:composite, name, children}
  end

  @doc "コンポジットに子を追加する"
  def add_child({:composite, name, children}, child) do
    {:composite, name, children ++ [child]}
  end

  @doc "ノードの値の合計を再帰的に計算する"
  def total({:leaf, _name, value}), do: value

  def total({:composite, _name, children}) do
    Enum.reduce(children, 0, fn child, acc -> acc + total(child) end)
  end

  @doc "ノードの数を再帰的にカウントする"
  def count({:leaf, _name, _value}), do: 1

  def count({:composite, _name, children}) do
    Enum.reduce(children, 1, fn child, acc -> acc + count(child) end)
  end

  @doc "ノードの名前を取得する"
  def name({:leaf, name, _value}), do: name
  def name({:composite, name, _children}), do: name

  @doc "木構造を文字列表現にする"
  def to_string_tree(node, indent \\ 0)

  def to_string_tree({:leaf, name, value}, indent) do
    String.duplicate("  ", indent) <> "#{name}: #{value}"
  end

  def to_string_tree({:composite, name, children}, indent) do
    prefix = String.duplicate("  ", indent)
    header = prefix <> "+ #{name}"
    child_strs = Enum.map(children, &to_string_tree(&1, indent + 1))
    Enum.join([header | child_strs], "\n")
  end
end
